### PR TITLE
HPCC-23364 Off-node storage

### PR DESCRIPTION
**Pull request by jakesmith**
Whilst continuing to use DFS and published groups, detected
containerized/cloud setup and use pseudo groups populated with
localhost.

Also move pv/pvc to be shared (not per Thor).

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review.

NB: I will open a separate JIRA to investigate what changes might be needed for DFU.

File: dockerfiles/hpcc/values.yaml [hunk 1]
```diff
@@ -5,6 +5,12 @@ global:
   
   dllserver:
     storageSize: 3Gi
+  
+  dataStorage:
+    storageSize: 100G
```
Comment (richardkchapman) [inline dockerfiles/hpcc/values.yaml]: This may be too large for some minikube systems

File: dockerfiles/hpcc/templates/roxie.yaml [hunk 2]
```diff
@@ -72,7 +75,8 @@ spec:
                 "--numChannels={{ $numChannels }}",
                 "--serverPorts={{ template "hpcc.utils.portListToCommas" $roxie.ports }}", 
                 "--topologyServers={{ $toponame }}:{{ $roxie.topoport }}",
-                "--localSlave=false"
+                "--localSlave=false", 
+                "--resolveLocally=false"
```
Comment (richardkchapman) [inline dockerfiles/hpcc/templates/roxie.yaml]: Need a similar change in localroxie.yaml

File: thorlcr/mfilemanager/thmfilemanager.cpp [hunk 3]
```diff
@@ -374,6 +381,19 @@ class CFileManager : public CSimpleInterface, implements IThorFileManager
 
     IFileDescriptor *create(CJobBase &job, const char *logicalName, StringArray &groupNames, IArrayOf<IGroup> &groups, bool overwriteok, unsigned helperFlags=0, bool nonLocalIndex=false, unsigned restrictedWidth=0)
     {
+        if (isCloud())
```
Comment (richardkchapman) [inline thorlcr/mfilemanager/thmfilemanager.cpp]: Would #ifdef _CONTAINERIZED be more sensible?

**Comment by richardkchapman** [comment]
@jakesmith A few minor comments

File: thorlcr/mfilemanager/thmfilemanager.cpp (hunk 3 above)
Comment (jakesmith) [reply thorlcr/mfilemanager/thmfilemanager.cpp]: I think I'll leave this as is for now.
It should really be a runtime isNas() I think, where normally true if _CONTAINERIZED, but in future if we support bare metal, me may want attached storage so same container with isNas()==false.

File: dockerfiles/hpcc/templates/dllserver.yaml [hunk 4]
```diff
@@ -7,9 +7,9 @@ metadata:
 spec:
   storageClassName: manual
   capacity:
-    storage: 10Gi
+    storage: {{ .Values.global.dllserver.storageSize }}
```
Comment (jakesmith) [inline dockerfiles/hpcc/templates/dllserver.yaml]: @richardkchapman - not directly related - but fixed.

File: dockerfiles/hpcc/templates/roxie.yaml (hunk 2 above)
Comment (jakesmith) [reply dockerfiles/hpcc/templates/roxie.yaml]: Added

**Comment by jakesmith** [comment]
@richardkchapman have made the small changes above and squashed in.

**Comment by jakesmith** [comment]
@richardkchapman - meant to say, I looked at removing storageSize storage.capacity, but it's required afaics:
```
Capacity
Generally, a PV will have a specific storage capacity. 
```
from : https://kubernetes.io/docs/concepts/storage/persistent-volumes/


Please review/merge.

File: dockerfiles/hpcc/templates/dllserver.yaml (hunk 4 above)
Comment (richardkchapman) [reply dockerfiles/hpcc/templates/dllserver.yaml]: Not 100% sure this is right. There's a subtle distinction between the volume size when setting up a local mount and the size of a pv request. Do we always want them to match?

**Comment by richardkchapman** [comment]
From the link you mention, I see:

Writing Portable Configuration

If you’re writing configuration templates or examples that run on a wide range of clusters and need persistent storage, it is recommended that you use the following pattern:
Include PersistentVolumeClaim objects in your bundle of config (alongside Deployments, ConfigMaps, etc).
Do not include PersistentVolume objects in the config, since the user instantiating the config may not have permission to create PersistentVolumes.
Give the user the option of providing a storage class name when instantiating the template.
If the user provides a storage class name, put that value into the persistentVolumeClaim.storageClassName field. This will cause the PVC to match the right storage class if the cluster has StorageClasses enabled by the admin.
If the user does not provide a storage class name, leave the persistentVolumeClaim.storageClassName field as nil. This will cause a PV to be automatically provisioned for the user with the default StorageClass in the cluster. Many cluster environments have a default StorageClass installed, or administrators can create their own default StorageClass.
In your tooling, watch for PVCs that are not getting bound after some time and surface this to the user, as this may indicate that the cluster has no dynamic storage support (in which case the user should create a matching PV) or the cluster has no storage system (in which case the user cannot deploy config requiring PVCs).

which looks like it could be relevant to us. Probably a topic for another Jira though

File: dockerfiles/hpcc/templates/dllserver.yaml [hunk 5]
```diff
@@ -7,9 +7,9 @@ metadata:
...
   accessModes:
     - ReadWriteMany
   hostPath:
-    path: {{ printf "%s/queries" $.Values.global.hostMountPath }}
+    path: {{ printf "%s/queries" $.Values.global.dataStorage.hostMountPath }}
```
Comment (richardkchapman) [inline dockerfiles/hpcc/templates/dllserver.yaml]: Dubious. dataStorage and dllServer are separate global sections, referring to separate storage needs. Should this say dllserver.hostMountPath?

File: dockerfiles/hpcc/values.yaml [hunk 6]
```diff
@@ -5,6 +5,12 @@ global:
   
   dllserver:
     storageSize: 3Gi
+  
+  dataStorage:
+    storageSize: 1G
```
Comment (richardkchapman) [inline dockerfiles/hpcc/values.yaml]: Let's be consistent about Gi vs G

**Comment by richardkchapman** [comment]
@jakesmith Some minor quivvles/questions, which would not stop me merging but which I'd like to talk about first

**Pull request merged**